### ISSUE TITLE
support for time dependent data sets

### ISSUE THREAD

**Issue by svater** [pyugrid/pyugrid#135]
The CF convention allows for time dependency, which should be also allowed in a UGRID file. There are even examples in the ugrid-conventions for this. If a variable is time dependent, an additional (unlimited) time dimension and variable has to be added in the file and the variable becomes two-dimensional with the first dimension being the time dimension and the second one being the dimension based on its location (face, node, edge). 

However, to my knowledge and according to my tests pyugrid does not support this, yet. I would like to introduce it and have already made some tests.

My suggestion would be to add an additional time attribute to the uvar class, which is None for time independent data and has the time vector for time dependent data. The data itself would become a two-dimensional numpy array with the first dimension for time. This has to be considered when the data is loaded and its consistency is checked. We might also want to save the time unit.

If anyone has already worked on this issue or if you have any thoughts about this, please comment. If you are fine with my envisioned layout I would like to make a concrete suggestion.

**Comment by svater** [comment]
Just for reference, I just saw that #28 already refers to this topic.

**Comment by ocefpaf** [comment]
@svater I am not really sure of what you are proposing here.  I have to admit that I don't pay much attention to `Uvar` as I am mostly interested in the `UGrid` functionality. I use other objects (raw `netCDf4-python` variables, `iris.cube`  and `xray` objects) to deal with the non-grid variables while using `Ugrid` to parse the unstructured grid for me.

However, @rsignell-usgs and @ChrisBarker-NOAA can comment if such implementation is needed.  Maybe the `UGrid` class might need something like this, to deal with grids that change with time, but I am not familiar with grids like that to comment.

**Comment by jay-hennen** [comment]
This is an issue that I'm currently dealing with in the code I'm writing for pygnome. `pyugrid` and `pysgrid` variables and grid both have no concept of time. However, one primary operation in our use case is interpolating the variable in time, before then interpolating in space. Currently the time interpolation is being implemented 'one level' up in the program. A container class `VectorField` has a time module that can interpolate in time, and a `pys/ugrid` that then does grid interpolation along with other functionality.

I'm not sure variable time interpolation belongs in a module meant to specialize in grid-related operations and representation. Time varying grids, on the other hand, might be an eventuality.

**Comment by ChrisBarker-NOAA** [comment]
I"ve come to the conclusion that yes, we need to handle time and z, once way or another. If you look at the conversation in #28, we were hoping to punt to Iris (or maybe now xarray) for that, but:

1) As much of the IRIS team likes the ides of supporting UGRIDs, they haven't done anything on it for a couple years -- and the IRIS code base is too painful for me to try to figure out how to do it.

2) A spoke with Stephan Hoyer (of xarray) about this last year at SciPy, and he was quite explicit: "xray will never use py_ugrid". He suggested that py_ugrid uses xarray instead.

So that leaves us where we are now, with py_ugrid only able to handle the horizontal spatial component, and users need to deal with the raw data arrays themselves, or use an odd mixture of IRIS or xarray and py_ugrid to do what they need to do. BUt this does not give us a friendly API that I think we all want.

So yes, we want UVars to do *something* with time (and the vertical dimension). However, where I think this should go, rather is to have an GridVar object that keeps track of time and the vertical, and uses one of: py_ugrid, py_srid, py_rect_grid (yet to be written)  under the hood to handle the horizontal dimensions.

So yes, this will mean re-writing some of what IRIS and xarray already do (could it use xarray???) but to get nice integration of the various grid types with a single API, I think that's the only way to go.

**Comment by ocefpaf** [comment]
> I've come to the conclusion that yes, we need to handle time and z, once way or another.

Not sure if we need to do this at the grid handling level. I guess that I am being thick headed thinking about `pysgrid` and `pyugrid` as grid handling tools only.

> IRIS code base is too painful for me to try to figure out how to do it.

As someone that contributed to iris I agree with your assessment, but we don't need to get `pyugrid` into `iris`.  I use both tools when working with CF-/UGRID-compliant files in production. Same is true for `xarray`. Maybe, if one day We think we am writing too much code and/or awkwardly loading too much redundant data, we might want to consider writing a library to wrap both at a higher level.  But I still advocate the idea of smaller and simpler libraries for specific jobs that can be easily re-used by others.

> "xray will never use py_ugrid"

I understand that.  The `xarray` library has an awkward support for non-regular grids because of the `pandas` inheritance. Same partially true for `iris`, but all the `CoordinateMultiDimError` in the code with a message `Not Implemented` gives me some hope.

Maybe the question is does the UGrid operations that we want fit into the `xarray.DataArray` or in the `iris.cube.Cube`? What are those operations? I cannot say what others do, but I use `pyugrid` to get easy access to the nodes, edges, and faces to build the matplotlib `tri` object.  Then I do the rest using that object (including finding nearest elements, which is blazing fast in matplotlib BTW.)

Again, if `pysgrid` and `pyugrid` become lightweight packages it won't matter if `iris` or `xarray` will use them.  We can just create our own combinations and solve the problem at hand.

PS: As APIRUS shows the real issue is to define all the problems we need/want to solve. I think that what @ayan-usgs did in pysgrid was the best approach for this issue.  He created the SGRID convention parser, `pysgrid`, and then added processing functions external to the object.  Only once we start using those functions we'll realize what needs to be folded back into the main `sgrid` object as part of the grid parser and what should remain external.  For example, the most common operation is the get the variables at the center for post-processing and plotting, so it is clear now that we need a method to get any variable at the center.

**Comment by jay-hennen** [comment]
> I cannot say what others do, but I use pyugrid to get easy access to the nodes, edges, and faces to build the matplotlib tri object. Then I do the rest using that object (including finding nearest elements, which is blazing fast in matplotlib BTW.)

Just would like to mention that finding the nearest node to a point (`locate_nodes()`) and locating the face which contains a point (`locate_faces') are available in pyugrid right now. I assume your use case goes past just those two functions, however. :)

> > I've come to the conclusion that yes, we need to handle time and z, once way or another.

> Not sure if we need to do this at the grid handling level. I guess that I am being thick headed thinking about pysgrid and pyugrid as grid handling tools only.

I hear you about pysgrid/pyugrid being grid handling tools, and I think the only good argument for a grid to be time-aware is if it changes in time, which isn't clear if we'll ever support. As for Z, you could argue it's technically a property of the grid and therefore should be handled at that level, eventually. Nobody really wants to be the first one to touch it though, and there's enough to do figuring out what we want for 2D support anyway.

One thing to point out however is that although time doesn't make sense for the grid, it can make sense for `UVar` and `SGridVariable` to be time aware, since interpolating between two time slices is a pretty common use case, and it's very common for time to be a dimension for grid variables. Other concrete applications escape me right now, but I suspect they're out there.

**Comment by ocefpaf** [comment]
> Just would like to mention that finding the nearest node to a point (locate_nodes()) and locating the face which contains a point (`locate_faces') are available in pyugrid right now. I assume your use case goes past just those two functions, however. :)

Thanks. (Actually those two are more than what I need :grin:)

> As for Z, you could argue it's technically a property of the grid and therefore should be handled at that level, eventually. Nobody really wants to be the first one to touch it though, and there's enough to do figuring out what we want for 2D support anyway.

Agreed. Not to mention that, once we compute z from the non-dimension form, we will be in trouble with a big array that depends on `eta(x, y, z, t)`. The best solution I found so far to deal with computed z is iris BTW.

> One thing to point out however is that although time doesn't make sense for the grid, it can make sense for `UVar` and `SGridVariable` to be time aware, since interpolating between two time slices is a pretty common use case, and it's very common for time to be a dimension for grid variables. Other concrete applications escape me right now, but I suspect they're out there.

What I am about to say is my personal opinion and by no means is what I will enforce (but I will suggest) in `py{u,s}grid`: there will be many functionalities and dependencies in a good `UVar` that it is worth splitting it from `pyugrid`. I believe that modules would evolve faster if this possible `UVar` module used `pyugrid` instead of being part of its core.

**Comment by ChrisBarker-NOAA** [comment]
I'm going to point out a little irony here -- on the one hand we are advocating splitting things up into multiple packages that each only do one thing -- makes some sense. On the other hand, we don't "need" to put in certain functionality, because it already exists in Monster Monolithic projects like Matplotlib. OK, I've said that :-)

As it happens, we only just discovered that MPL includes a "find which triangle you are in", which does look to be pretty good -- we'll probably do some performance testing, comparisons, etc against cell_tree2d. I did know that there is a fair bit of triangular grid logic in MPL, and when that was begin added I encouraged the author to split it out from MPL as a stand alone triangle mesh handling library, but he didn't want to have the extra hassle of maintaining a package, and keeping versions and APIs in sync, etc. There is a lot of overhead in additional packages (less than there was, but still...).

And the fact that stuff like point location is buried in MPL is why I didn't find it long ago...

While I've been advocating putting more into py_ugrid and py_sgrid, we did make celltree_2d a stand alone package, and we've been doing the same thing with a bunch of our other stuff (py_gd, unit_conversion, etc.) so far, there hasn't been much (any?) use of those outside of our tools, so it's just package maintaining overhead for no benefit...

Al this is to say that there is no one right answer about how much to bundle things up into a large package vs. separate packages. Each decision needs to made based on specific needs.

So on to the specifics:

@ocefpaf wrote: """n py{u,s}grid: there will be many functionalities and dependencies in a good UVar that it is worth splitting it from pyugrid. I believe that modules would evolve faster if this possible UVar module used pyugrid instead of being part of its core."""

That's exactly the direction I think we're headed -- a variable on a Ugrid and a Variable on a Sgrid (or a regular grid, for that matter) is essentially the same thing, except that you need to navigate the horizontal dimensions differently. And time and vertical are all handled the same way.

In the end, what I'd really like is an abstraction for a field of information -- you should be able to do all sorts of stuff with it without knowing or caring what kind of grid it's on. 

So this points to a Variable object of some sort that has a grid object that it can query for grid-related info. Which all comes back to the APIRUS idea of py_ugrid and py_sgrid sharing the same API.

I'll go post on APIRUS about this -- in the end, I suspect it's better that we have one package that has the Variable object, and y_ugrid, and py_sgrid, and py_rectgrid, and... rather than 5 or six interrelated packages (the MPL idea...), but we should start with keep them separate and see how it pans out.

NOTE: point location and interpolation do belong in the *_grid projects -- how that is done is intimately associated with the nature of the grid.

@hetland: bringing you in, because this is really about the intersection of py_ugrid and py_sgrid.

**Comment by ChrisBarker-NOAA** [comment]
@svater: AS you see, it's unclear right now where time handling code should go, but yes, we do need it, so concrete suggestions are all good.

**Comment by ChrisBarker-NOAA** [comment]
NOTE: 
""" I use other objects (raw netCDf4-python variables, iris.cube and xray objects) to deal with the non-grid variables while using Ugrid to parse the unstructured grid for me."""

While this works fine for some (most) cases, it only works if your data are all in netcdf files (or openDAP servers compliant with the netcdf lib) -- but it's one level of abstraction too low -- what if the data are in other formats, or, you're going in the other direction, and the data are come from your app, and are only in memory, etc?  This is all why I want py_ugrid to be an abstraction for the entire dataset, not just the grid -- that abstraction can be something that *uses* py_ugrid, rather than *being* py_ugrid, but it would be a gdo thing to have.

Another note here -- when you want to deal with more than one variable, or even worse two related variables (like U and V for velocity), it really gets ugly...

**Comment by ocefpaf** [comment]
> I'm going to point out a little irony here -- on the one hand we are advocating splitting things up into multiple packages that each only do one thing -- makes some sense. On the other hand, we don't "need" to put in certain functionality, because it already exists in Monster Monolithic projects like Matplotlib. OK, I've said that :-)

I think is a mistake.  I would love to see the tri-stuff out of Matplotib.  (Together with the in_polygon-stuff.)

> I did know that there is a fair bit of triangular grid logic in MPL.

I only realized that after SciPy. @tacaswell made a quick comment that made me curious and I went to check it out.  I guess it is not well advertised. (I pinged him here hoping he will comment some more on the tri-stuff.)

> This is all why I want py_ugrid to be an abstraction for the entire dataset.

Iris is an abstraction for CF and not netCDF.  It can deal with a variety of files and even custom loading. I am unsure how flexible `xarray` is, but I do not that netCDF is not the only option. (They have a pynio backend that reads a ton of stuff.)

**Comment by ChrisBarker-NOAA** [comment]
> > This is all why I want py_ugrid to be an abstraction for the entire dataset.

> Iris is an abstraction for CF and not netCDF. It can deal with a variety of files and even custom loading.
> I am unsure how flexible xarray is, but I do not that netCDF is not the only option. (They have a pynio
> backend that reads a ton of stuff.)

Exactly :-) -- which is why it would all work out great if IRIS or xarray *used* py_ugrid, etc. in order to take advantage of that abstraction, it all needs to work together.

Also -- probably a documentation issue more than anything, but creating a set of IRIS cubes from scratch is challenging...

**Comment by tacaswell** [comment]
The mpl triangulation code is used internally and I do not think we intend to expose it as part of the public api (so we can break it when ever we want :wink: )

@ianthomas23 is the correct person to ask about that code.

**Comment by ocefpaf** [comment]
> I do not think we intend to expose it as part of the public api (so we can break it when ever we want :wink: )

I would not mind re-writing my stuff every now and then to be able to use that without installing matplotlib. That code is awesome!

**Comment by ChrisBarker-NOAA** [comment]
thanks @tacaswell -- though we actually have the opposite problem -- we may want to use the triangulation code -- so we need it to be a stable API....

I know @ianthomas23 didn't want to spilt it out as a separate library , and this is a good reason why -- if you keep it internal, you can mess with the internal API all you want.

NOTE: this is a reason to keep py_ugrid, py_sgrid, and whatever variable objects it may use coupled.

**Comment by ianthomas23** [comment]
Hello folks.  I can confirm that, as a couple of you have commented, I haven't split out any of the mpl triangular grid code into a separate package because I don't want the hassle of leading a standalone project.  But also it is not intended to be a triangle mesh **handling** library, it is a triangle mesh **plotting** library and hence sensibly resides within a wider plotting library.  Admittedly the tri finder, interpolation and mesh refinement have uses outside of just plotting, but they were motivated by plotting.

It would be great if there was a sufficiently good and comprehensive python tri grid library that became a community standard.  If such a library wanted to use any of the mpl tri grid code they would be welcome to use it, and I expect  that I would provide support if it was wanted.  But I have no interest in being a core developer of such a project.

Of course, anyone can use any of the mpl source code as they wish within the limits of the licence, so it could be incorporated into other projects rather than forcing those other projects to depend on mpl.

The internal API, in particular that between the C++ and python tri grid code, is indeed private because we reserve the right to change it whenever we wish.  In reality I don't recall it changing at all, so it is no less stable than our public API.

**Comment by ocefpaf** [comment]
Thanks for the explanation @ianthomas23. Even thought I would love to see that code as a stand alone library out there I understand your points.

**Comment by hetland** [comment]
One solution might be to just steal the code right out of MPL, and use it
in our code. A bit silly, but reduces the dependency, and does not get
broken when MPL changes.

Also, there is quite a bit of Triangle stuff in scipy now. I expect qhull
is a great way to do things. But that is an even more stringent dependency.

BTW what is the code in MPL? Last I checked it had moved. Any examples?

On Sun, Mar 6, 2016 at 12:58 PM, Filipe <notifications@github.com> wrote:

> Thanks for the explanation @ianthomas23 <https://github.com/ianthomas23>.
> Even thought I would love to see that code as a stand alone library out
> there I understand your points.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyugrid/pyugrid/issues/135#issuecomment-192962936>.
>



-- 
Prof. Rob Hetland
Texas A&M Univ.  –  Dept. of Oceanography
http://pong.tamu.edu/~rob

**Comment by ocefpaf** [comment]
> One solution might be to just steal the code right out of MPL, and use it
in our code. A bit silly, but reduces the dependency, and does not get
broken when MPL changes.

I thought about that, but right now I cannot do it. (I just added to my never ending TODO list.)

> BTW what is the code in MPL? Last I checked it had moved. Any examples?

I made a very simple blog post using pyugrid and mpl [here](https://ocefpaf.github.io/python4oceanographers/blog/2015/07/20/pyugrid/).

**Comment by rhattersley** [comment]
Jumping off from one of @ChrisBarker-NOAA's comments:

> So this points to a Variable object of some sort that has a grid object that it can query for grid-related info.

That's where I'd start too. It's always been my hand-wavy vision for this stuff, since before Iris even made it onto GitHub.

> Exactly :-) -- which is why it would all work out great if IRIS or xarray used py_ugrid, etc. in order to take advantage of that abstraction, it all needs to work together.

Personally I'd love to see Iris making use of a grid abstraction. We took a very first step in that direction when @cbcunc visited the Met Office, but unfortunately project constraints have prevented us taking this further as yet.

> Also -- probably a documentation issue more than anything, but creating a set of IRIS cubes from scratch is challenging...

I don't want to hijack a pyugrid issue, but it does overlap so ... suggestions for improvements to Iris are _very_ welcome. The Met Office core devs probably won't have time to lead that kind of effort for at least six months, but we can still be responsive to proposals and PRs, plus there are other non-MO core devs and that list can expand.

**Comment by svater** [comment]
It would be nice if we could keep with the topic of this issue. In my opinion the discussion about features of MPL to be included into py_ugrid or not is a bit off-topic. 

Also, I think time dependent grids is a completely different topic. Although it is surely of great concern - actually, we have such grids, which are adapted in each time step and a triangle is devided into two or the other way around. However, we solve this issue by using a different NetCDF file for each time step, at the moment.

I am referring to simply a time series of a grid variable, e.g. 

`double Mesh2_waterlevel(time,nMesh2_face)`

where only the values change in time. In the current state such variables cannot be loaded with py_ugrid and this is a very bad state for new people trying to use py_ugrid in my opinion. I think most will abandon this software, because they think it simply does not work correctly. There could even be time dependent and independent variables (such as waterlevel and bathymetry) in one file. So either, one should deprecate such broken functionality and describe how to work with the variables on the grid in the documentation, or put this into a working state.

I like the idea that there should be not too many dependencies in py-ugrid. I had a short look at xarray and iris, and these packages have probably a great functionality. However, I am not sure of how much use they are for the average user. Also, I have only very limited time, which means I would stick with a simple solution for the time being.

I think it would be great to just have one small library like py_ugrid is at the moment, which enables users to work with UGRID files in a simple way in python. Even some simple plotting functionality would be good for that. Interpolation between time levels might be another neat functionality to be added.

Btw., what do you mean by z-dependency? Do you refer to a 3D-layered mesh topology (as described in the UGRID conventions) or something else? The case of a 3D-layered mesh topology seems to not being well described in the conventions for the variables, but it would be probably handled similar to time.

**Comment by jay-hennen** [comment]
>  In the current state such variables cannot be loaded with py_ugrid and this is a very bad state for new people trying to use py_ugrid in my opinion. I think most will abandon this software, because they think it simply does not work correctly.

There's code for this, in the `expanded_uvar` branch, which I think also contains some interpolation stuff and other improvements. I'll have to go back and check, it's been a while. The code isn't in `master` yet for vague reasons relating to uncertain design intent for the project, as far as I can tell. (aka idk why)

It should let you 'wrap' a netCDF `Variable` in a `UVar` and then use standard `[]` format to slice it like you would in netCDF.

`u = pyugrid.UVar('u', 'node', dataset['u'])`

**Comment by ChrisBarker-NOAA** [comment]
@svater:

Yes, gotten a bit off-topic -- maybe a new issue is two is in order. But the key question here, which is on topic, is what functionality should be included in py_ugrid (what libs are used to get that is another matter...)

> I am referring to simply a time series of a grid variable, e.g.

 `double Mesh2_waterlevel(time,nMesh2_face)`

Absolutely -- that should "Just work" at least I think so -- and UVars have been in PyUgrid from before it did anything useful at all -- but we all had high hopes for integartion with IRIS (and then xarray), so they don't get a lot of attention.

See @jay-hennen's notes about what he's added to UVar.

NOTE: a key issue here is perspective:

do you want to load up a "Variable" object of some sort -- say representing water lever, and have the grid be transparent (yes, I think so...)

Or do you want to be able to have one object that is the entire (or a subset) of the dataset? -- i.e. the grid AND the data on it. This more or less mirrors what is in a typical netcdf file of model results. (yes, I think we want this, too, and that's what UGrid supports, at least a little).

However, as the VAriable object needs to support time, and yes, layers model Z, and other stuff that is in common with other grid types, we should probably pull that out to a project that can be shared.
